_drafts/conviva-dag-process-time-prod-issue-draft.md

```diff
@@ -0,0 +1,160 @@
+
+<div class="aside">This is not unlike how [snapshot isolation](https://jepsen.io/consistency/models/snapshot-isolation) works in databases with multi-version concurrency control. The purpose is, of course, different but there are overlaps in the mechanism.<br/><br/></div>
+
+`ArcSwap` accomplishes this with a [thread-local epoch counter to track "debt"](https://github.com/vorner/arc-swap/blob/master/src/debt/list.rs#L335) which avoids cache contention issues that crop up when updating a shared read counter.
+
+A new version of the data is swapped in using the [standard `cmp_xchg`](https://github.com/vorner/arc-swap/blob/master/src/strategy/hybrid.rs#L207) operation. This marks the beginning of a new epoch, but the data associated with the old epoch isn't cleaned up until all "debt" is paid off, that is until all readers of the previous epoch have finished. Once the readers from the previous epoch have finished, the previous epoch is "reclaimed".
+
+```text
+  [Core 1]             [Core 2]             [Core 3]
+     |                   |                   |
+     | load()            | load()            | load()
+     |                   |                   |
+     v                   v                   v
+
+[Thread-Local Epoch 1] [Thread-Local Epoch 2] [Thread-Local Epoch 3]
+      (read guard)           (read guard)         (read guard)
+
+      ┌─────────────────────────────┐
+      │        ArcSwap<T>          │
+      │ ┌────────────────────────┐ │
+      │ │ Arc<T>: Current value  │ │ <── atomic ptr (no cache bouncing)
+      │ └────────────────────────┘ │
+      └─────────────────────────────┘
+
+     Writer swaps in new Arc<T> using atomic store()
+     └── Old Arc<T> placed into deferred queue
+         └── Only dropped when all read guards released
+
+✅ No shared counters for read
+✅ No cache line bouncing
+✅ Readers are wait-free and isolated
+```
+
+The big difference between a concurrent hash map and `ArcSwap` is that `ArcSwap` requires swapping out the entirety of the underlying data with every write but trades this off with very cheap reads. Writes don't even have to wait for all readers to finish since a new epoch is created with the new version of data.
+
+Hash maps on the other hand allow updating invidual portions of data in the hash map but this is where it becomes important that we have an *almost* read-only scenario because the additional overhead of writes with `ArcSwap` is worth paying here since reads are faster.
+
+
+## Conclusion
+Given that we had a situation which was almost read-only, the overhead of a concurrent hash map was not suitable since we had no use case for frequent, granular updates. Trading that for `ArcSwap`, which is a specialized `AtomicRef`, something that is designed for occasional swaps where the entire ref is updated is much better fit for our use case.
```

Comment:
is _a_ much better fit

Reply:
addressed by 940b1d3bb898a8f79e24a6fafd6c9dc69d4ab0c7